### PR TITLE
feat/adata: implement shell and owner_at_index getter functions for AppendOnly Data and minor fixes

### PR DESCRIPTION
**Pull request by Yoga07**

### CODE REVIEW

File: src/append_only_data.rs [hunk 1]
```diff
@@ -547,6 +568,57 @@ impl SeqAppendOnlyData<PubPermissions> {
             },
         }
     }
+
+    pub fn shell(&self, index: u64) -> Result<Self> {
+
+        if self.entry_index() < index {
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: I would suggest replacing this with
```suggestion
        if index > self.entry_index() {
```
to make the intent more clear (checking for overflows).

File: src/append_only_data.rs [hunk 2]
```diff
@@ -547,6 +568,57 @@ impl SeqAppendOnlyData<PubPermissions> {
             },
         }
     }
+
+    pub fn shell(&self, index: u64) -> Result<Self> {
+
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: There's extraneous whitespace, could you please run rustfmt on it?

**Comment by Yoga07** [comment]
> I wonder if we can avoid code duplication on this.
> Would it be possible to move the `shell` impl into the `impl_appendable_data` macro? I think the shell should be mostly the same for all types of `AppendOnlyData`.

Ah I did try writing the function inside the trait. It required the `Sized` trait to be implemented to the structs to return a `Self`.

**Comment by nbaksalyar** [comment]
>I did try writing the function inside the trait. It required the Sized trait to be implemented to the structs to return a Self.

It does not necessarily have to be defined in the trait @Yoga07.
So what I mean is we have this macro `impl_appendable_data` which implements trait methods for different AppendOnlyData types, but it can also be used to implement struct-level methods.

So it can be just this, after the `pub struct $flavour<P>` block inside the macro:

```rust
impl<P> $flavour<P>
where
  P: Permissions + std::hash::Hash + Clone,
{
  pub fn shell(&self, index: u64) -> Result<Self> {
    ..
  }
}
```

File: src/lib.rs [hunk 3]
```diff
@@ -66,7 +66,7 @@ mod request;
 mod response;
 
 pub use append_only_data::{
-    Address as ADataAddress, AppendOnlyData, Index as ADataIndex, Indices as ADataIndices,
+    Action, Address as ADataAddress, AppendOnlyData, Index as ADataIndex, Indices as ADataIndices,
```
Comment (nbaksalyar) [inline src/lib.rs]: I think this should be
```suggestion
    Action as ADataAction, Address as ADataAddress, AppendOnlyData, Index as ADataIndex, Indices as ADataIndices,
```
to be not confused with `MDataAction`.

File: src/append_only_data.rs [hunk 4]
```diff
@@ -265,11 +265,21 @@ impl Permissions for PubPermissions {
...
+    pub public_key: PublicKey,
     /// The current index of the data when this ownership change happened
-    data_index: u64,
+    pub data_index: u64,
     /// The current index of the permissions when this ownership change happened
-    permissions_index: u64,
+    pub permissions_index: u64,
+}
+
+impl Owner {
```
Comment (Fraser999) [inline src/append_only_data.rs]: Is there a need for this impl since the members are public now?

File: src/append_only_data.rs [hunk 5]
```diff
@@ -357,14 +370,49 @@ macro_rules! impl_appendable_data {
...
+                    if owner.data_index() <= index {
+                        owner_vec.push(owner.clone());
+                    }
+                }
+
+                Ok(Self {
+                    inner: AppendOnly {
+                        address: self.inner.address,
+                        data: Vec::new(),
+                        permissions: perm_vec.clone(),
```
Comment (Fraser999) [inline src/append_only_data.rs]: I don't think we need to clone here or on the line below.

File: src/append_only_data.rs [hunk 6]
```diff
@@ -417,6 +465,14 @@ macro_rules! impl_appendable_data {
                 }
             }
 
+            fn fetch_owner_at_index(&self, owners_index: u64) -> Option<&Owner> {
+                if self.inner.owners.len() >= owners_index as usize {
```
Comment (Fraser999) [inline src/append_only_data.rs]: This could just be `self.inner.owners.get(owners_index as usize)`

File: src/append_only_data.rs [hunk 7]
```diff
@@ -357,14 +370,49 @@ macro_rules! impl_appendable_data {
...
+        impl<P> $flavour<P>
+        where
+            P: Permissions + std::hash::Hash + Clone,
+        {
+            pub fn shell(&self, index: u64) -> Result<Self> {
+                if index > self.entry_index() {
+                    return Err(Error::NoSuchEntry);
+                }
+
+                let mut perm_vec: Vec<P> = Vec::new();
```
Comment (Fraser999) [inline src/append_only_data.rs]: We could probably just use `filter` here?  Or `take_while` if the indices are always held in increasing order?  E.g.
```rust
let permissions = self
    .inner
    .permissions
    .iter()
    .filter(|perm| perm.data_index() <= index)
    .cloned()
    .collect();
let owners = self
    .inner
    .owners
    .iter()
    .filter(|owner| owner.data_index <= index)
    .cloned()
    .collect();
```

File: src/append_only_data.rs [hunk 8]
```diff
@@ -357,14 +360,50 @@ macro_rules! impl_appendable_data {
...
+                    .iter()
+                    .filter(|owner| owner.data_index <= index)
+                    .cloned()
+                    .collect();
+
+                Ok(Self {
+                    inner: AppendOnly {
+                        address: self.inner.address,
+                        data: Vec::new(),
+                        permissions: permissions,
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: Can be just
```suggestion
                        permissions,
```

File: src/append_only_data.rs [hunk 9]
```diff
@@ -357,14 +360,50 @@ macro_rules! impl_appendable_data {
...
+                        owners: owners,
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: same here
```suggestion
                        owners,
```

**Pull request merged**